File: AlgoExpert/JavaScript/BinaryTrees/branchSums.js
// write a function that takes in a Binary Tree and returns a list of its branch sums ordered from leftmost branch sum to rightmost branch sum

// a branch sum is the sum of all the values in a binary tree branch

class BinaryTree {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

function branchSums(root) {
    let sum = []; // will contain the sum of each branch as [a, b, c]

    calculator(root, 0, sum);

    return sum;
};

function calculator(node, currentSum, sums) {
    if (node === null) return 0;
    let newSum = currentSum + node.value;

    if (!node.left && !node.right) {
        sums.push(newSum);
    }

    calculator(node.left, newSum, sums);
    calculator(node.right, newSum, sums);
}

File: AlgoExpert/Python/BST/branch_sums.py
# write a function that takes in a Binary Tree and returns a list of its branch sums ordered from leftmost branch sum to rightmost branch sum

# a branch sum is the sum of all values in a BT branch.

# NOTES
# we need to go down each branch and add the values for each node
# there are three types of traversals
# we can focus on in-order traversal

# PSEUDOCODE
# create a sums variable containing an empty list
# create a helper function

# helper function:
# should take the root, starting sum = 0, and sums
# current_sum = starting sum + node.value
# solve this recursively so,
# if root is None:
# return
# if node.left is None and node.right is None:
# sums.append(node)

# helper function(node.left, current_sum, sums)
# helper function(node.right, current_sum, sums)


class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def branchSums(root):
    sums = []

    return branchSumsHelper(root, 0, sums)


def branchSumsHelper(node, starting_sum, sums):
    if node is not None:
        return

    current_sum = starting_sum + node.value

    if node.left is None and node.right is None:
        sums.append(current_sum)
        return

    branchSumsHelper(node.left, current_sum, sums)
    branchSumsHelper(node.right, current_sum, sums)
